kernel/cls_allocator/src/lib.rs

```diff
@@ -0,0 +1,77 @@
+#![no_std]
+
+extern crate alloc;
+
+use alloc::{sync::Arc, vec::Vec};
+
+use cpu::CpuId;
+use crate_metadata::{LoadedSection, StrongSectionRef};
+use local_storage_initializer::{ClsDataImage, ClsInitializer, LocalStorageInitializerError};
+use sync_spin::SpinMutex;
+
+static CLS_INITIALIZER: SpinMutex<ClsInitializer> = SpinMutex::new(ClsInitializer::new());
+static CLS_SECTIONS: SpinMutex<Vec<(CpuId, ClsDataImage)>> = SpinMutex::new(Vec::new());
+
+/// Adds a CLS section with a pre-determined offset to the global CLS
+/// initializer.
+///
+/// The CLS register will not be updated until either [`reload`] or
+/// [`reload_current_cpu`] is called.
+pub fn add_static_section(
+    section: LoadedSection,
+    offset: usize,
+    total_static_storage_size: usize,
+) -> Result<StrongSectionRef, LocalStorageInitializerError> {
+    CLS_INITIALIZER
+        .lock()
+        .add_existing_static_section(section, offset, total_static_storage_size)
+}
+
+/// Adds a dynamic CLS section to the global CLS initializer.
+///
+/// The CLS register will not be updated until either [`reload`] or
+/// [`reload_current_cpu`] is called.
+pub fn add_dynamic_section(
+    section: LoadedSection,
+    alignment: usize,
+) -> Result<(usize, Arc<LoadedSection>), LocalStorageInitializerError> {
+    CLS_INITIALIZER
+        .lock()
+        .add_new_dynamic_section(section, alignment)
+}
+
+/// Generates a new data image for the current CPU and sets the CLS register
+/// accordingly.
+pub fn reload_current_cpu() {
+    let current_cpu = cpu::current_cpu();
+
+    let mut data = CLS_INITIALIZER.lock().get_data();
+
+    let mut sections = CLS_SECTIONS.lock();
```

Comment:
in reading this function, i'm getting mildly confused by the terminology used and how it differs from TLS. 

This `CLS_SECTIONS` isn't a set of sections, right? It's a set of data images, according to your type def of it, so i think it's name should be changed to reflect that.

My understanding of the terminology is as follows (and tbh i'm likely not 100% consistent with this, but we should be moving forward):
* a TLS/CLS data image is the _starting_ value that gets populated by the content of `.tdata`/`.cls` sections. The term image is fitting, as we consider it read-only from the perspective of executing code... though it will be changed when dynamically loading a new section.
* a TLS/CLS data region or area is an instance of a data image that was allocated for and given to a specific task or CPU. 
* a TLS/CLS section is actually a symbol, but yea, that one is clear.

Reply:
> This CLS_SECTIONS isn't a set of sections, right?

Yea that's a mistake on my part.

> a TLS/CLS data image is the starting value that gets populated by the content of .tdata/.cls sections. The term image is fitting, as we consider it read-only from the perspective of executing code... though it will be changed when dynamically loading a new section.

I don't think that's true. When we set an image as the current base (for both TLS/CLS) it can be modified by the process/cpu implicitly through TLS/CLS accesses. The `Tls/ClsInitializer` is what stores the starting value.

> a TLS/CLS data region or area is an instance of a data image that was allocated for and given to a specific task or CPU.

No, that's a data image.

Reply:
hmm, did i mix them up? That's how I recall reading about TLS images. 

Perhaps I created that set of terminology and it's not canonical? My above description matches how I use that language in Theseus's TLS-related crates. I mean, overall it's just a semantic choice to differentiate between "data image" and "data area/region", but we just need to be consistent. 

If you have a better suggestion for clear terminology here, I'm totally open to it. In my mind, an "image" is a sort of snapshot of something, so I wouldn't expect it to be modified in general usage, but rather just serve as a starting point. Kinda COW-style sematics.

Reply:
Looks like the best name for the image is "TLS initial[izer] image".  I don't see a canonical name for the data area instance owned by a specific task or CPU.

> I don't think that's true. When we set an image as the current base (for both TLS/CLS) it can be modified by the process/cpu implicitly through TLS/CLS accesses. The `Tls/ClsInitializer` is what stores the starting value.

Sorry, forgot to respond. What I said was correct -- the behavior currently on `theseus_main` is to call [`TlsInitializer::get_data()`](https://github.com/theseus-os/Theseus/blob/27429e57120be7b6725101fde4582dfd48b936d0/kernel/tls_initializer/src/lib.rs#L250C49-L251C1) to obtain a new instance of the TLS data image for a newly-spawned task, and the FSBASE register is set to point to that new instance, not the single instance cached in the `TlsInitializer`.


Looking at that linked code, I see the confusion. That `get_data()` function returns a `TlsDataImage`, which I refer to as a "TLS data area" -- it is precisely at that point (the call to `get_data()`) that I consider it to have become a TLS data area (or whatever we want to name it), as it is now specific to a task and can be modified by regular `%fs:[OFFSET]` access instructions.

Reply:
The [Fuchsia docs](https://fuchsia.dev/fuchsia-src/development/kernel/threads/tls) refer to each thread's instance as a "TLS Region", which itself is composed of multiple "TLS Block"s. 

We don't care about the distinction between various TLS Blocks since we only support the Local Exec TLS model. 

Let's go with `TLS Initial Image` and `TLS Region`.... later, in a future PR. We'll fix it up then for the sake of consistency.

Reply:
So, long story short, this indicates that `CLS_SECTIONS` should be called `CLS_REGIONS`.